url/url.cpp

```diff
@@ -51,4 +101,1341 @@ std::string blob_url_create(Origin const &origin) {
+    icu_init();
+
+    std::string ascii_domain;
+    icu::StringByteSink<std::string> tmp{&ascii_domain};
+
+    icu::IDNAInfo inf;
+    UErrorCode err = U_ZERO_ERROR;
+
+    std::uint32_t opts = UIDNA_NONTRANSITIONAL_TO_ASCII | UIDNA_CHECK_BIDI | UIDNA_CHECK_CONTEXTJ;
+
+    if (be_strict) {
+        opts |= UIDNA_USE_STD3_RULES;
+    }
+
+    auto uts = icu::IDNA::createUTS46Instance(opts, err);
+
+    if (U_FAILURE(err)) {
+        spdlog::info("Failed to create UTS46 instance, error {}; idna data probably missing from icu",
+                static_cast<std::int64_t>(err));
+
+        return std::nullopt;
+    }
+
+    err = U_ZERO_ERROR;
+
+    uts->nameToASCII_UTF8(domain, tmp, inf, err);
+
+    delete uts;
+
+    std::uint32_t proc_err = inf.getErrors();
+
+    // icu doesn't offer a flag to disable VerifyDnsLength or CheckHyphens, so just ignore those failures
+    proc_err &= ~UIDNA_ERROR_LEADING_HYPHEN;
+    proc_err &= ~UIDNA_ERROR_TRAILING_HYPHEN;
+    proc_err &= ~UIDNA_ERROR_HYPHEN_3_4;
+
+    if (!be_strict) {
+        proc_err &= ~UIDNA_ERROR_LABEL_TOO_LONG;
+        proc_err &= ~UIDNA_ERROR_DOMAIN_NAME_TOO_LONG;
+    }
+
+    // If domain or any label is empty, proc_err should contain UIDNA_ERROR_EMPTY_LABEL
+    if (U_FAILURE(err) || proc_err != 0) {
+        validation_error(ValidationError::DomainToASCII);
+
+        return std::nullopt;
+    }
+
+    return ascii_domain;
+}
+
+// https://url.spec.whatwg.org/#start-with-a-windows-drive-letter
+bool UrlParser::starts_with_windows_drive_letter(std::string_view input) {
+    if (input.size() >= 2) {
+        if (util::is_alpha(input[0]) && (input[1] == ':' || input[1] == '|')) {
+            if (input.size() == 2 || input[2] == '/' || input[2] == '\\' || input[2] == '?' || input[2] == '#') {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+// https://url.spec.whatwg.org/#shorten-a-urls-path
+void UrlParser::shorten_url_path(Url &url) {
+    assert(!std::holds_alternative<std::string>(url.path));
+
+    if (url.scheme == "file" && std::get<1>(url.path).size() == 1
+            && util::is_normal_windows_drive_letter(std::get<1>(url.path)[0])) {
+        return;
+    }
+
+    std::get<1>(url.path).pop_back();
+}
+
+// https://url.spec.whatwg.org/#concept-host-parser
+std::optional<Host> UrlParser::parse_host(std::string_view input, std::optional<bool> is_not_special) {
+    assert(!input.empty());
+
+    if (input.starts_with("[")) {
+        if (!input.ends_with("]")) {
+            validation_error(ValidationError::IPv6Unclosed);
+
+            return std::nullopt;
+        }
+
+        input.remove_prefix(1);
+        input.remove_suffix(1);
+
+        std::optional<std::array<std::uint16_t, 8>> addr = parse_ipv6(input);
+
+        if (!addr.has_value()) {
+            return std::nullopt;
+        }
+
+        return Host{HostType::Ip6Addr, *addr};
+    }
+
+    if (*is_not_special) {
```

Comment:
This optional is never checked, so this is not safe. Did you mean to just pass a bool and check if that was false?

Reply:
That parameter has a default value of `false`, if you look at the declaration in the header. That said, yeah, this should just be a bool. I'm not sure why the spec calls it an optional parameter.